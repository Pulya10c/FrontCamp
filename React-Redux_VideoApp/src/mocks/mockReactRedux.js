const createMockMapStateToProps = mapStateToProps => {
  if (!mapStateToProps) {
    return mapStateToProps;
  }

  return (state, ownProps) => {
    return {
      props: mapStateToProps(state, ownProps)
    };
  };
};

const getDispatchFunctionFormShorthandObject = actionsObj => dispatch => {
  return Object.keys(actionsObj).reduce((props, actionKey) => {
    return {
      ...props,
      [actionKey]: function() {
        return dispatch(actionsObj[actionKey].apply(this, arguments));
      }
    };
  }, {});
};

const createMockMapDispatchToProps = mapDispatchToProps => {
  if (!mapDispatchToProps) {
    return mapDispatchToProps;
  }

  const mappingFunction =
    typeof mapDispatchToProps === 'function'
      ? mapDispatchToProps
      : getDispatchFunctionFormShorthandObject(mapDispatchToProps);

  return (dispatch, ownProps) => {
    dispatch = dispatch || jest.fn(action => action);

    return {
      props: mappingFunction(dispatch, ownProps),
      dispatch
    };
  };
};

const createMockMergeProps = mergeProps => {
  if (!mergeProps) {
    return mergeProps;
  }

  return (stateProps, dispatchProps, ownProps) => {
    dispatchProps = dispatchProps || { dispatch: jest.fn(action => action) };

    return {
      props: mergeProps(stateProps, dispatchProps, ownProps),
      dispatch: dispatchProps.dispatch
    };
  };
};

export default {
  connect: (
    mapStateToProps,
    mapDispatchToProps,
    mergeProps
  ) => ConnectedComponent => ({
    mapStateToProps: createMockMapStateToProps(mapStateToProps),
    mapDispatchToProps: createMockMapDispatchToProps(mapDispatchToProps),
    mergeProps: createMockMergeProps(mergeProps),
    ConnectedComponent
  })
};
