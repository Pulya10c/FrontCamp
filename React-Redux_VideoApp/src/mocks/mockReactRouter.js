import React from 'react';

const Route = () => <div className="mock-react-router-route"></div>;

Route.propTypes = {};

const Router = () => <div className="mock-react-router-router"></div>;

const Consumer = () => <div className="mock-react-router-consumer"></div>;

const Switch = () => <div className="mock-react-router-switch"></div>;

const Redirect = () => <div className="mock-react-router-redirect"></div>;

export default {
  Route,
  Router,
  Switch,
  Redirect,
  __RouterContext: {
    Consumer
  },
  withRouter: ConnectedComponent => ({
    ConnectedComponent
  })
};
