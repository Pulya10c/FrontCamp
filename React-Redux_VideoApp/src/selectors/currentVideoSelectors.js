import { createSelector } from 'reselect';
import get from 'lodash/get';

export const currentVideoSelector = state => get(state, 'currentVideo');

export const currentVideoGenresSelector = createSelector(
  [currentVideoSelector],
  currentVideo => get(currentVideo, 'genres', [])
);

export const filmPosterPathSelector = createSelector(
  [currentVideoSelector],
  currentVideo => get(currentVideo, 'poster_path') // add default img
);

export const filmTitleSelector = createSelector(
  [currentVideoSelector],
  currentVideo => get(currentVideo, 'title')
);

export const filmRatingSelector = createSelector(
  [currentVideoSelector],
  currentVideo => get(currentVideo, 'vote_average')
);

export const filmReleaseDateSelector = createSelector(
  [currentVideoSelector],
  currentVideo => get(currentVideo, 'release_date')
);

export const filmOverviewSelector = createSelector(
  [currentVideoSelector],
  currentVideo => get(currentVideo, 'overview')
);

//  currentVideo: {
//     id: 351286,
//     title: 'Jurassic World: Fallen Kingdom',
//     tagline: 'Life finds a way',
//     vote_average: 0,
//     vote_count: 28,
//     release_date: '2018-06-01',
//     poster_path: 'https://image.tmdb.org/t/p/w500/ln6d5Okr6VK5vfQVobJTiYxeD0l.jpg',
//     overview: 'A volcanic eruption threatens the remaining dinosaurs on the island of Isla Nublar, where the creatures have freely roamed for several years after the demise of an animal theme park known as Jurassic World. Claire Dearing, the former park manager, has now founded the Dinosaur Protection Group, an organization dedicated to protecting the dinosaurs. To help with her cause, Claire has recruited Owen Grady, a former dinosaur trainer who worked at the park, to prevent the extinction of the dinosaurs once again.',
//     budget: 0,
//     revenue: 0,
//     genres: [
//       'Action',
//       'Adventure',
//       'Drama',
//       'Science Fiction',
//       'Thriller'
//     ],
//     runtime: null
//   }
