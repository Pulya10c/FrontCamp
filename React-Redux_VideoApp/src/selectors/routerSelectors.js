import { createSelector } from 'reselect'
import get from 'lodash/get';
import { PATHS } from 'src/constants'

export const routerSelector = state => get(state, 'router');

export const locationSelector = createSelector(
  [routerSelector],
  router => get(router, 'location')
);

export const routerSearchSelector = createSelector(
  [locationSelector],
  location => get(location, 'search')
);

export const pathNameSelector = createSelector(
  [locationSelector],
  location => get(location, 'pathname')
);


export const isVideoListPath = createSelector(
  [locationSelector],
  location => get(location, 'pathname') === PATHS.VIDEO_LIST_PAGE
);

