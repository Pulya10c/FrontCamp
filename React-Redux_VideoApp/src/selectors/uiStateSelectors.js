import { createSelector } from 'reselect'
import get from 'lodash/get';

export const uiStateSelector = state => get(state, 'uiState');

export const fetchingSelector = createSelector(
  [ uiStateSelector ],
  uiState => get(uiState, 'fetching', false)
);

export const errorSelector = createSelector(
  [ uiStateSelector ],
  uiState => get(uiState, 'error', null)
);

