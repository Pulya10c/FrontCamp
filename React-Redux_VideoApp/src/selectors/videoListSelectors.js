import { createSelector } from 'reselect'
import get from 'lodash/get';

export const videoListSelector = state => get(state, 'videoList');

export const filterConfigSelector = createSelector(
  [videoListSelector],
  videoList => get(videoList, 'filterConfig')
);

export const searchSelector = createSelector(
  [filterConfigSelector],
  filterConfig => get(filterConfig, 'search')
);

export const searchBySelector = createSelector(
  [filterConfigSelector],
  filterConfig => get(filterConfig, 'searchBy')
);
