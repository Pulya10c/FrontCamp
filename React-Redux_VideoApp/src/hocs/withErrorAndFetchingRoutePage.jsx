import React from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import { withRouter } from "react-router";

import Spinner from "components/Spinner";
import { fetchingSelector, errorSelector } from "selectors/uiStateSelectors";

const withErrorAndFetchingRoutePage = PageComponent => {
  const mapStateToProps = state => ({
    fetching: fetchingSelector(state),
    error: errorSelector(state)
  });

  const InitialPage = ({ error, fetching, dispatch, ...routeProps }) => {
    return (
      <>
        {fetching && <Spinner />}
        {error && <span>Нешта здохла... =\</span>}
        <PageComponent {...routeProps} />
      </>
    );
  };

  return withRouter(connect(mapStateToProps, null)(InitialPage));
};

export default withErrorAndFetchingRoutePage;
