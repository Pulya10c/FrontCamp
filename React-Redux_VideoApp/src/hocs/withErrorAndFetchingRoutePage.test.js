import React from 'react';
import { shallow } from 'enzyme';
import mockReactRedux from 'mocks/mockReactRedux';
import mockReactRouter from 'mocks/mockReactRouter';
import * as uiStateSelectors from 'selectors/uiStateSelectors';
import withErrorAndFetchingRoutePage from './withErrorAndFetchingRoutePage';

jest.mock('react-router', () => mockReactRouter);
jest.mock('react-redux', () => mockReactRedux);
uiStateSelectors.fetchingSelector = jest.fn(() => false);
uiStateSelectors.errorSelector = jest.fn(() => false);

describe('Component with withErrorAndFetchingRoutePage', () => {
  let wrapper, TargetPage, ReduxConnectedComponent, TargetPageContainer;

  const PageComponent = props => <div {...props}>Page Component</div>;
  TargetPage = withErrorAndFetchingRoutePage(PageComponent);
  ReduxConnectedComponent = TargetPage.ConnectedComponent;
  TargetPageContainer = ReduxConnectedComponent.ConnectedComponent;

  beforeEach(() => {
    wrapper = shallow(<TargetPageContainer />);
  });

  it('should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
