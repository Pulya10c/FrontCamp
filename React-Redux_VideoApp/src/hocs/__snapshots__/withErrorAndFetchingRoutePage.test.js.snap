// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component with withErrorAndFetchingRoutePage should render correctly 1`] = `
<Fragment>
  <PageComponent />
</Fragment>
`;
